File: android/ilds-design-system/src/main/kotlin/com/icicilombard/ilds/components/IldsSelectionButton.kt
package com.icicilombard.ilds.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.Immutable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.selected
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.icicilombard.ilds.tokens.IldsTokens

// Figma set 14776:1685 — mirrors web/SelectionButton.tsx + lib/ilds_selection_button.dart.

enum class IldsSelectionButtonSize { Small, Medium, Large }
enum class IldsSelectionButtonVariant { LabelOnly, LabelWithSuffix, IconOnly }

@Composable
fun IldsSelectionButton(
    label: String,
    isSelected: Boolean,
    onPress: () -> Unit,
    modifier: Modifier = Modifier,
    size: IldsSelectionButtonSize = IldsSelectionButtonSize.Medium,
    variant: IldsSelectionButtonVariant = IldsSelectionButtonVariant.LabelOnly,
    isDisabled: Boolean = false,
    suffixIcon: (@Composable () -> Unit)? = null,
    semanticLabel: String? = null,
) {
    val metrics = remember(size) { IldsSelectionButtonMetrics(size) }
    val interactionSource = remember { MutableInteractionSource() }
    val isPressed by interactionSource.collectIsPressedAsState()
    val colors = remember(isSelected, isDisabled, isPressed) {
        IldsSelectionButtonColors.resolve(isSelected, isDisabled, isPressed)
    }
    val iconOnly = variant == IldsSelectionButtonVariant.IconOnly
    val showSuffix = (variant == IldsSelectionButtonVariant.LabelWithSuffix || iconOnly) && suffixIcon != null
    val a11yLabel = semanticLabel ?: label

    Surface(
        onClick = onPress,
        modifier = modifier
            .semantics {
                contentDescription = a11yLabel
                selected = isSelected
                role = Role.Button
            },
        enabled = !isDisabled,
        interactionSource = interactionSource,
        shape = RoundedCornerShape(IldsTokens.radiusMedium),
        color = colors.background,
        border = BorderStroke(colors.borderWidth, colors.border),
    ) {
        Row(
            modifier = Modifier
                .defaultMinSize(minHeight = metrics.height)
                .padding(horizontal = metrics.horizontalPadding),
            horizontalArrangement = Arrangement.spacedBy(IldsTokens.sp4),
            verticalAlignment = Alignment.CenterVertically,
        ) {
            if (!iconOnly) {
                Text(
                    text = label,
                    style = metrics.textStyle,
                    color = colors.text,
                    maxLines = 1,
                )
            }
            if (showSuffix) {
                suffixIcon?.invoke()
            }
        }
    }
}

@Immutable
private data class IldsSelectionButtonColors(
    val background: Color,
    val border: Color,
    val borderWidth: Dp,
    val text: Color,
) {
    companion object {
        fun resolve(isSelected: Boolean, isDisabled: Boolean, isPressed: Boolean): IldsSelectionButtonColors {
            if (isDisabled) {
                return IldsSelectionButtonColors(
                    IldsTokens.neutralCoolgray50,
                    IldsTokens.neutralCoolgray100,
                    1.dp,
                    IldsTokens.neutralCoolgray300,
                )
            }
            if (isSelected) {
                return IldsSelectionButtonColors(
                    IldsTokens.primaryOrange50,
                    IldsTokens.primaryOrange500,
                    2.dp,
                    IldsTokens.primaryOrange500,
                )
            }
            return IldsSelectionButtonColors(
                background = if (isPressed) IldsTokens.neutralCoolgray100 else IldsTokens.globalWhite000,
                border = if (isPressed) IldsTokens.neutralCoolgray300 else IldsTokens.neutralCoolgray200,
                borderWidth = 1.dp,
                text = if (isPressed) IldsTokens.neutralCoolgray900 else IldsTokens.neutralCoolgray600,
            )
        }
    }
}

@Immutable
private data class IldsSelectionButtonMetrics(
    val height: Dp,
    val horizontalPadding: Dp,
    val textStyle: TextStyle,
) {
    constructor(size: IldsSelectionButtonSize) : this(
        height = when (size) {
            IldsSelectionButtonSize.Small -> 32.dp
            IldsSelectionButtonSize.Medium -> 40.dp
            IldsSelectionButtonSize.Large -> 48.dp
        },
        horizontalPadding = when (size) {
            IldsSelectionButtonSize.Small -> IldsTokens.sp8
            IldsSelectionButtonSize.Medium -> IldsTokens.sp12
            IldsSelectionButtonSize.Large -> IldsTokens.sp16
        },
        textStyle = TextStyle(
            fontSize = when (size) {
                IldsSelectionButtonSize.Small -> IldsTokens.fontSize12
                IldsSelectionButtonSize.Medium -> IldsTokens.fontSize14
                IldsSelectionButtonSize.Large -> IldsTokens.fontSize16
            },
            lineHeight = when (size) {
                IldsSelectionButtonSize.Small -> 16.sp
                IldsSelectionButtonSize.Medium -> 18.sp
                IldsSelectionButtonSize.Large -> 20.sp
            },
            fontWeight = IldsTokens.fontWeightMedium,
        ),
    )
}

File: ios/Sources/ILDSDesignSystem/IldsSelectionButton.swift
import ILDSTokens
import SwiftUI

// Mirrors lib/ilds_selection_button.dart + web SelectionButton.tsx. TODO(web/desktop): hover states — mobile-first defer.

public enum IldsSelectionButtonSize: Sendable {
    case small, medium, large
}

public enum IldsSelectionButtonVariant: Sendable {
    case labelOnly, labelWithSuffix, iconOnly
}

public struct IldsSelectionButton: View {
    private let label: String
    private let isSelected: Bool
    private let action: () -> Void
    private let suffixIcon: AnyView?
    private let variant: IldsSelectionButtonVariant
    private let size: IldsSelectionButtonSize
    private let isDisabled: Bool

    public init(
        _ label: String,
        isSelected: Bool,
        action: @escaping () -> Void,
        suffixIcon: (any View)? = nil,
        variant: IldsSelectionButtonVariant = .labelOnly,
        size: IldsSelectionButtonSize = .medium,
        isDisabled: Bool = false
    ) {
        self.label = label
        self.isSelected = isSelected
        self.action = action
        self.suffixIcon = suffixIcon.map { AnyView($0) }
        self.variant = variant
        self.size = size
        self.isDisabled = isDisabled
    }

    public var body: some View {
        Button(action: action) {
            IldsSelectionButtonLabel(
                label: label,
                isSelected: isSelected,
                suffixIcon: suffixIcon,
                variant: variant,
                size: size,
                isDisabled: isDisabled,
                isPressed: false,
                isFocused: false
            )
        }
        .buttonStyle(IldsSelectionButtonStyle(
            label: label,
            isSelected: isSelected,
            suffixIcon: suffixIcon,
            variant: variant,
            size: size,
            isDisabled: isDisabled
        ))
        .disabled(isDisabled)
        .accessibilityLabel(label)
        .accessibilityAddTraits(isSelected ? .isSelected : [])
    }
}

private struct IldsSelectionButtonLabel: View {
    let label: String
    let isSelected: Bool
    let suffixIcon: AnyView?
    let variant: IldsSelectionButtonVariant
    let size: IldsSelectionButtonSize
    let isDisabled: Bool
    let isPressed: Bool
    let isFocused: Bool

    var body: some View {
        let metrics = IldsSelectionButtonMetrics(size: size)
        let colors = IldsSelectionButtonColors.resolve(
            isSelected: isSelected,
            isDisabled: isDisabled,
            isPressed: isPressed,
            isFocused: isFocused
        )
        let iconOnly = variant == .iconOnly

        HStack(spacing: ILDSTokens.sp4) {
            if !iconOnly {
                Text(label)
                    .font(metrics.font)
                    .foregroundStyle(colors.text)
            }
            if variant == .labelWithSuffix, let suffixIcon {
                suffixIcon
                    .foregroundStyle(colors.text)
                    .frame(width: metrics.iconSize, height: metrics.iconSize)
            }
            if iconOnly {
                Group {
                    if let suffixIcon {
                        suffixIcon
                    } else {
                        Image(systemName: "checkmark")
                    }
                }
                .foregroundStyle(colors.text)
                .frame(width: metrics.iconSize, height: metrics.iconSize)
            }
        }
        .frame(height: metrics.height)
        .padding(.horizontal, metrics.horizontalPadding)
        .background(colors.background)
        .overlay {
            RoundedRectangle(cornerRadius: ILDSTokens.radiusMedium)
                .strokeBorder(colors.border, lineWidth: colors.borderWidth)
        }
        .clipShape(RoundedRectangle(cornerRadius: ILDSTokens.radiusMedium))
    }
}

private struct IldsSelectionButtonStyle: ButtonStyle {
    let label: String
    let isSelected: Bool
    let suffixIcon: AnyView?
    let variant: IldsSelectionButtonVariant
    let size: IldsSelectionButtonSize
    let isDisabled: Bool

    func makeBody(configuration: Configuration) -> some View {
        IldsSelectionButtonLabel(
            label: label,
            isSelected: isSelected,
            suffixIcon: suffixIcon,
            variant: variant,
            size: size,
            isDisabled: isDisabled,
            isPressed: configuration.isPressed && !isDisabled,
            isFocused: false
        )
    }
}

private struct IldsSelectionButtonMetrics {
    let height: CGFloat
    let horizontalPadding: CGFloat
    let font: Font
    let iconSize: CGFloat

    init(size: IldsSelectionButtonSize) {
        switch size {
        case .small:
            height = ILDSTokens.sp32
            horizontalPadding = ILDSTokens.sp8
            font = .system(size: ILDSTokens.fontSize12, weight: ILDSTokens.fontWeightMedium)
            iconSize = ILDSTokens.fontSize12
        case .medium:
            height = ILDSTokens.sp40
            horizontalPadding = ILDSTokens.sp12
            font = .system(size: ILDSTokens.fontSize14, weight: ILDSTokens.fontWeightMedium)
            iconSize = ILDSTokens.fontSize14
        case .large:
            height = ILDSTokens.sp48
            horizontalPadding = ILDSTokens.sp16
            font = .system(size: ILDSTokens.fontSize16, weight: ILDSTokens.fontWeightMedium)
            iconSize = ILDSTokens.fontSize16
        }
    }
}

private struct IldsSelectionButtonColors {
    let background: Color
    let text: Color
    let border: Color
    let borderWidth: CGFloat

    static func resolve(
        isSelected: Bool,
        isDisabled: Bool,
        isPressed: Bool,
        isFocused: Bool
    ) -> IldsSelectionButtonColors {
        let background: Color
        if isDisabled {
            background = ILDSTokens.neutralCoolgray50
        } else if isSelected {
            background = ILDSTokens.primaryOrange50
        } else if isPressed {
            background = ILDSTokens.neutralCoolgray100
        } else {
            background = ILDSTokens.globalWhite000
        }

        let text: Color
        if isDisabled {
            text = ILDSTokens.neutralCoolgray300
        } else if isSelected || isFocused {
            text = ILDSTokens.primaryOrange500
        } else if isPressed {
            text = ILDSTokens.neutralCoolgray900
        } else {
            text = ILDSTokens.neutralCoolgray600
        }

        let border: Color
        if isDisabled {
            border = ILDSTokens.neutralCoolgray100
        } else if isSelected || isFocused {
            border = ILDSTokens.primaryOrange500
        } else if isPressed {
            border = ILDSTokens.neutralCoolgray300
        } else {
            border = ILDSTokens.neutralCoolgray200
        }

        let borderWidth: CGFloat = isSelected || isFocused ? 2 : 1
        return IldsSelectionButtonColors(background: background, text: text, border: border, borderWidth: borderWidth)
    }
}
